WhoAreYou: add obscureText feature

Feichtmeier opened the pull request:
![ya](https://user-images.githubusercontent.com/15329494/148093116-69b9658c-f037-443c-86c3-d8b9b0aac9d9.gif)



Closes #589

jpnurmi commented on packages/ubuntu_wizard/lib/src/widgets/validated_form_field.dart:
```diff
@@ -56,6 +56,15 @@ class ValidatedFormField extends StatefulWidget {
   /// Enables or disables TextField (defaults to true)
   final bool enabled;
 
+  /// An optional custom icon for the show text button
+  final IconData? showTextIcon;
+
+  /// An optional custom icon for the show text button
+  final IconData? hideTextIcon;
+
+  /// Used when the show text button is pressed
+  final VoidCallback? onShowText;
+
```
For a generic widget such as `ValidatedFormField`, it would be nicer to allow any widget by exposing `InputDecoration.prefix`. The logic for showing and hiding the password should be in `_PasswordFormField`.

jpnurmi commented:
Thanks, looking good so far. I'd prefer to keep `ValidatedFormField` use case -agnostic and move the logic to `_PasswordFormField`, though.

To make the CI pass, you can run `melos generate` (or `flutter run build_runner build --delete-conflicting-outputs`) to generate up-to-date mocks. Last but not least, the change should be paired with appropriate tests. :wink:

Feichtmeier commented:
> Thanks, looking good so far. I'd prefer to keep `ValidatedFormField` use case -agnostic and move the logic to `_PasswordFormField`, though.
> 
> To make the CI pass, you can run `melos generate` (or `flutter run build_runner build --delete-conflicting-outputs`) to generate up-to-date mocks. Last but not least, the change should be paired with appropriate tests. wink

Thanks for the review, moved the code out of the validatedformfield and updated the test (i hope correctly?)

However none of the two commands worked for me. From which directory should I run them?

jpnurmi commented on packages/ubuntu_wizard/lib/src/widgets/validated_form_field.dart:
```diff
@@ -123,9 +131,9 @@ class _ValidatedFormFieldState extends State<ValidatedFormField> {
-      ),
+          labelText: widget.labelText,
+          helperText: widget.helperText,
+          suffixIcon: widget.suffixIcon),
```
can you add the missing comma after "widget.suffixIcon" to fix the indentation?

jpnurmi commented on packages/ubuntu_wizard/lib/src/widgets/validated_form_field.dart:
```diff
@@ -56,6 +56,12 @@ class ValidatedFormField extends StatefulWidget {
+  /// An optional Widget placed inside the text input.
+  final Widget? suffixIcon;
+
+  /// Used when the show text button is pressed
+  final VoidCallback? onShowText;
```
can be removed

jpnurmi commented on packages/ubuntu_desktop_installer/lib/pages/who_are_you/who_are_you_widgets.dart:
```diff
@@ -133,6 +135,17 @@ class _PasswordFormField extends StatelessWidget {
         final model = Provider.of<WhoAreYouModel>(context, listen: false);
         model.password = value;
       },
+      onShowText: () {
+        final model = Provider.of<WhoAreYouModel>(context, listen: false);
+        model.obscureText = !model.obscureText;
+      },
```
can be removed

jpnurmi commented:
> However none of the two commands worked for me. From which directory should I run them?

`flutter run build_runner build --delete-conflicting-outputs` needs to be run for a specific package. In this case, we need to run it in the `packages/ubuntu_desktop_installer` directory.

`melos generate` is a convenient and easy to remember script that runs the same command for all packages. It can be run from anywhere within the repo.

Feichtmeier commented:
I didn't have melos installed :) Installed now, thanks! and the command worked!

Melos also updated the tests for
ethernet and
connecttointernet
should I commit those generated mocks or only the ones related to this PR?

jpnurmi commented:
> should I commit those generated mocks or only the ones related to this PR?

Only the ones related to this PR please :)

jpnurmi commented:
The test seems to have failed in the CI. Did it pass locally?
```
01:48 +43: /home/runner/work/ubuntu-desktop-installer/ubuntu-desktop-installer/packages/ubuntu_desktop_installer/test/who_are_you_page_test.dart: obscure text                                         
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure object was thrown running a test:
  Expected: exactly one matching node in the widget tree
  Actual: _AncestorFinder:<zero widgets with type "IconButton" which is an ancestor of icon
"IconData(U+0E6BE)">
   Which: means none were found but one was expected

When the exception was thrown, this was the stack:
#4      main.<anonymous closure> (file:///home/runner/work/ubuntu-desktop-installer/ubuntu-desktop-installer/packages/ubuntu_desktop_installer/test/who_are_you_page_test.dart:226:5)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///home/runner/work/ubuntu-desktop-installer/ubuntu-desktop-installer/packages/ubuntu_desktop_installer/test/who_are_you_page_test.dart line 226
The test description was:
  obscure text
════════════════════════════════════════════════════════════════════════════════════════════════════

01:48 +43 -1: /home/runner/work/ubuntu-desktop-installer/ubuntu-desktop-installer/packages/ubuntu_desktop_installer/test/who_are_you_page_test.dart: obscure text [E]                                  
  Test failed. See exception logs above.
  The test description was: obscure text
```

jpnurmi reviewed (approved):
Thanks, LGTM :+1:
